src/main/java/org/opensearch/knn/index/vectorvalues/KNNVectorValuesIterator.java

```diff
@@ -185,4 +189,96 @@ public VectorValueExtractorStrategy getVectorExtractorStrategy() {
         }
     }
 
+    abstract class MergeSegmentVectorValuesIterator<T extends DocIdSetIterator, U> implements KNNVectorValuesIterator {
+
+        private DocIDMerger<KNNMergeVectorValues.KNNVectorValuesSub<T>> docIdMerger;
+        private final int liveDocs;
+        private int docId;
+        protected KNNMergeVectorValues.KNNVectorValuesSub<T> current;
+
+        private static final VectorValueExtractorStrategy VECTOR_VALUES_STRATEGY =
+            new VectorValueExtractorStrategy.MergeSegmentValuesExtractor();
+
+        MergeSegmentVectorValuesIterator(final List<KNNMergeVectorValues.KNNVectorValuesSub<T>> subs, final MergeState mergeState)
+            throws IOException {
+            this.docIdMerger = DocIDMerger.of(subs, mergeState.needsIndexSort);
+            int totalSize = 0;
+            for (KNNMergeVectorValues.KNNVectorValuesSub<T> sub : subs) {
+                totalSize += sub.liveDocs;
+            }
+            this.liveDocs = totalSize;
+            this.docId = -1;
+
+        }
+
+        @Override
+        public int docId() {
+            return docId;
+        }
+
+        @Override
+        public int advance(int docId) throws IOException {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public int nextDoc() throws IOException {
+            current = docIdMerger.next();
+            if (current == null) {
+                docId = NO_MORE_DOCS;
+            } else {
+                docId = current.mappedDocID;
+            }
+            return docId;
+        }
+
+        @Override
+        public DocIdSetIterator getDocIdSetIterator() {
+            // while we can get the values of current, this method is intended to be called once so it's much better to throw
+            // so Liskov-Substitution-principle is not violated unknowingly
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public long liveDocs() {
+            return liveDocs;
+        }
+
+        @Override
+        public VectorValueExtractorStrategy getVectorExtractorStrategy() {
+            return VECTOR_VALUES_STRATEGY;
+        }
+
+        public abstract U vectorValue() throws IOException;
+    }
+
+    class MergeFloat32VectorValuesIterator extends MergeSegmentVectorValuesIterator<FloatVectorValues, float[]> {
```

Comment:
Why we went with creating the iterator route here? rather we could have just extended the FloatVectorValues and ByteVectorValues. This would have avoided all these iterators and new strategy classes. 

Ref: https://github.com/apache/lucene/blob/branch_9x/lucene/core/src/java/org/apache/lucene/codecs/KnnVectorsWriter.java#L286

Reply:
Didn't realize its a DISI, changing it

We can take up the lucene contribution as a follow up

Reply:
> We can take up the lucene contribution as a follow up

Please create a gh issue and ref it here.